src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -175,14 +162,19 @@ private void CleanupOldPayloads(object? sender, EventArgs e)
             return t.Result;
         }
 
-        public Block? GetPayload(string payloadId)
+        public async ValueTask<Block?> GetPayload(string payloadId)
         {
-            if (_payloadStorage.TryRemove(payloadId, out BlockImprovementContext? blockContext))
+            if (_payloadStorage.TryRemove(payloadId, out IBlockImprovementContext? blockContext))
             {
-                // TODO: We can wait here a bit ~500ms for the block production to complete.
-                // Client software MAY stop the corresponding build process after serving this call.
-                blockContext.Cancel();
-                return blockContext.CurrentBestBlock;
+                using (blockContext)
+                {
+                    if (!blockContext.ImprovementTask.IsCompleted)
+                    {
+                        await Task.WhenAny(blockContext.ImprovementTask, Task.Delay(500));
```

Comment:
QQ: We removed blockContext.Cancel? is this okay?

Reply:
Yes, moved to dispose pattern and added using here.